### ISSUE TITLE
Add warning in Dataset.shuffle API for possible data leakage

### ISSUE THREAD

**Issue by HaoyuYang59** [tensorflow/tensorflow#59279]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Documentation Feature Request

### Have you reproduced the bug with TF nightly?

No

### Source

source

### Tensorflow Version

2.9,2.10,2.11

### Custom Code

Yes

### Current Behaviour?

```shell
Hi Tensorflow Team,

Could you please add a pitfall warning in the Docs of "tf.data.Dataset" API, shuffle method about the "reshuffle_each_iteration" might lead to validation data leakage? 

It's only very recently that I learned using shuffle method followed by take/skip methods to generate train/test/validation sets could lead to val/test data leaking into training set (because the full dataset is shuffled before split for each epoch).
```

### Standalone code to reproduce the issue

```shell
dataset = tf.data.Dataset.from_tensor_slices((features, labels))
dataset = dataset.shuffle(buffer_size=BUFFER_SIZE)
dataset = dataset.batch(BATCH_SIZE)

train_dataset = dataset.take(TRAIN_BATCH_SIZE)
val_dataset = dataset.skip(TRAIN_BATCH_SIZE)

model.fit(train_dataset, validation_data=val_dataset, batch_size=BATCH_SIZE, epochs=EPOCHS)
```

### Relevant log output

_No response_</details>

**Comment by gaikwadrahul8** [comment]
Hi, @HaoyuYang59 

Thank you so much for noticing the data leakage issue with `dataset.shuffle` with `reshuffle_each_iteration=True` option and I was able to replicate data leakage issue with some random dataset values, you can find gist file [here](https://colab.research.google.com/gist/gaikwadrahul8/b54a085807b105be1a91c4f8e6957590/-59279.ipynb) and to aware of this data leakage issue to other users with `dataset.shuffle` with `reshuffle_each_iteration=True` option I'll raise PR to add some note or warning on our [official documentation ](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#shuffle)very soon and I'll update the same here, Thank you!

**Comment by gaikwadrahul8** [comment]
Hi, @HaoyuYang59 

I have submitted PR #59310 to update either note or warning about data leakage issue w.r.t `dataset.shuffle `with `reshuffle_each_iteration=True` option on our official documentation and it will be taken care once PR got merged.
Thank you!

**Comment by HaoyuYang59** [comment]
Hi @gaikwadrahul8,

Thanks for devoting attention to this issue. And great to know it would be addressed shortly. Hope it could help more avoid this pitfall.

Regards,
Haoyu

**Comment by gaikwadrahul8** [comment]
Hi, @HaoyuYang59 

I have re-opened this issue to take care of underlying issue for data leakage with `tf.data.Dataset.shuffle `with option `(reshuffle_each_iteration=True)` as per this [comment](https://github.com/tensorflow/tensorflow/pull/59310#issuecomment-1400718726) and concerned team will take care of this issue very soon. Thank you!

**Comment by HaoyuYang59** [comment]
Hi @gaikwadrahul8,

Thanks for your attention on this issue, and I deeply appreciate it.

Yes I have been following that commit as well. 

The reason I required a "warning" instead of a "BUG fix" is that from my understanding this behaviour is **more or less "intended by the developer" but "unexpected by the user"**. As a user I didn't expect the dataset to be shuffled each epoch when the `shuffle` method is call only once (and more importantly that `reshuffle_each_iteration=True` is default). And this answer on stackoverflow [here](https://stackoverflow.com/questions/48213766/split-a-dataset-created-by-tensorflow-dataset-api-in-to-train-and-test) using the same method got **58** upvote so far.

Or perhaps the `take` and `skip` methods should be prohibited from being used after `shuffle (reshuffle_each_iteration=True)`? 

Also I noticed later version of TensorFlow seems to have another `tf.keras.utils.split_dataset` method which prevented data leakage [here](https://www.tensorflow.org/api_docs/python/tf/keras/utils/split_dataset).

Thanks for your time.

**Comment by sachinprasadhs** [comment]
Hi, Could you please provide more debug information on validation data leakage. Thanks!

**Comment by HaoyuYang59** [comment]
@sachinprasadhs Hi thanks for your reply. @gaikwadrahul8 has generously helped me with an example code [here](https://colab.research.google.com/gist/gaikwadrahul8/b54a085807b105be1a91c4f8e6957590/-59279.ipynb) to recreate this data leakage issue.

More specifically using `dataset.take()` after `dataset.shuffle(reshuffle_each_iteration=True)` (where `reshuffle_each_iteration=True` is default) would lead to the whole dataset being shuffled and split after each epoch, as a result validation data would leak into training set.

However I fully understand this is NOT is bug, and from my understanding it's something works as intended by developer however is unexpected by user. As a user I didn't expect the dataset to be shuffled after each epoch when `shuffle` is called only once.

Also I believe many others are making the same mistake as [this answer](https://stackoverflow.com/questions/48213766/split-a-dataset-created-by-tensorflow-dataset-api-in-to-train-and-test) on StackOverflow has got 58 upvotes so far, and unfortunately would lead to data leakage.

Thanks for sending the guide however I'm afraid it didn't mention splitting the dataset into training and validation sets.

Regards,
Haoyu

**Comment by KaleabTessera** [comment]
Was this ever fixed?

**Comment by HaoyuYang59** [comment]
@sachinprasadhs @gaikwadrahul8 Could you please have a look at this issue? Thanks.

**Comment by deepGio** [comment]
Same behavior here. 

There are many references to Tensorflow Dataset API using skip() and take() pattern for splitting training, validation, and test sets (two methods that do exactly what is supposed to be done in this case with no mentioned warnings). One of the most used function is like the following:

`def get_dataset_partitions_tf(ds, ds_size, train_split=0.8, val_split=0.1, test_split=0.1, shuffle=True, shuffle_size=10000):
    assert (train_split + test_split + val_split) == 1
    
    if shuffle:
        # Specify seed to always have the same split distribution between runs
        ds = ds.shuffle(shuffle_size, seed=12)
    
    train_size = int(train_split * ds_size)
    val_size = int(val_split * ds_size)
    
    train_ds = ds.take(train_size)    
    val_ds = ds.skip(train_size).take(val_size)
    test_ds = ds.skip(train_size).skip(val_size)
    
    return train_ds, val_ds, test_ds`

As occurred to me, using shuffle (a very common case) before sets split defenitely leads to awful and dangerously silent data leakage, compromising all performance metrics.
I think it is crucial to alert the user with a warning and fix the behavior as soon as possible.

**Comment by aazuspan** [comment]
+1 to the suggestion of adding a warning if possible, or at least clearly documenting this behavior. It is very easy to accidentally leak data and inflate model performance with the default behavior, and based on the StackOverflow answers mentioned above, it's a common misunderstanding that's still being widely shared.

**Comment by DanielYang59** [comment]
Welcome to the new year! Unfortunately, it seems that the attention of maintainers is yet to be directed towards this persistent issue in TensorFlow. The problem persists even in TF2.14 without any warnings in the documentation.

If anyone is interested, we may consider drafting a PR to tackle this issue ourselves. (@deepGio, @aazuspan), especially since I lack experience contributing to a project as extensive as TF. Currently, I see several aspects to address:

- Adding a warning in the documentation is a considerate approach, though it may not be the most effective, as not all users  read the documentation.

- The default argument of `reshuffle_each_iteration=None` in dataset.shuffle is misleading, as discussed earlier, where **None is interpreted as True**. This could be clarified in the documentation, but it might not reach all users.

- Changing the source code to set the default value of `reshuffle_each_iteration = False` could be highly effective in preventing data leakage. However, it might be too drastic, potentially impacting the results of many users, especially those unaware of the data leakage issue and make them confused.

- Triggering a warning for dataset.shuffle calls might be a viable option, although it could be bothersome if the warning is generated for every call. It might be more appropriate to **issue a warning only when model.fit() is called on a dataset using the problematic "shuffle-take-skip" combination** (with reshuffle=True). Unfortunately, I am not familiar with how to achieve this, so any guidance or advice would be greatly appreciated.

In the meantime, it would be invaluable to receive insights from other members of the TensorFlow community as well. Any advice from Tensorflowers would be highly appreciated too. (@gaikwadrahul8, @aaudiber, @sachinprasadhs)

**Comment by aaudiber** [comment]
There are some straightforward changes that could help here:

- Add a warning about reshuffle behavior, especially in regard to take/skip. 
- Change the default value from `None` to `True`. AFAICT this would have no effect on behavior, but makes the API much less misleading.

I'm happy to LGTM a PR for these changes.

I like the idea of issuing a warning when the shuffle + take/skip pattern is observed. The shuffle dataset would need to start tracking which datasets are derived from it, so that it can complain when there is both a skip and take with the same argument. To do this we could track `outputs` for each dataset stage, similar to how we currently track [inputs](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/ops/dataset_ops.py#L414-L418). I think we would want to store outputs as weak references to avoid GC issues due to reference cycles. This would be a complex enough change that we should weight the risk of regressions against the benefit of the warning. I feel like the warning is justified given the popularity of the pattern in https://stackoverflow.com/questions/48213766/split-a-dataset-created-by-tensorflow-dataset-api-in-to-train-and-test. @wilsingosti what do you think?

**Comment by DanielYang59** [comment]
Thanks a lot for your comment @aaudiber. I could try make these changes but I'm not confident about how to implement a proper detection of the `reshuffle + take/skip` pattern. I might go ahead for a PR and looking forward to see further discussion there!

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by DanielYang59** [comment]
I'm currently not working on this so perhaps I would just close it (not sure if this got fixed in later versions though)
